ConstActualValueUsageAnalyzer: #256: Allow Assert on Enum.Member

manfred-brands opened the pull request:
This fixes #256 
It will always fire if actual is a literal as that is the classic error case.
It will however not fire if actual is a constant field and expected is also a constant.
This allows validating values of named enumeration values and constants.

Dreamescaper commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,16 +26,56 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
+                        .Where(e => e != null)
+                        .FirstOrDefault();
+                }
+            }
```
nit: could you extract `GetExpectedExpression` to separate method?

manfred-brands commented:
Addressed review request.
Noticed that negative numbers were not considered to be constants. Added code to detect constant literal expression in general.
Added special exemption for '.Empty' in cases like: Assert.AreEqual(actual, string.Empty)

manfred-brands commented:
Added a fixup for the .Empty case to allow: Assert.That(SomeCollection.Empty, Is.Empty)

Dreamescaper commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +26,91 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
+            bool IsEmpty(ExpressionSyntax expression)
+            {
+                return expression is MemberAccessExpressionSyntax memberAccessExpression &&
+                    memberAccessExpression.Name.Identifier.Text == "Empty";
```
Personally I don't mind considering `string.Empty` as const for this analyzer, but I'm not sure it applies to any `Empty` property possible.
I would limit it by strings only.

manfred-brands replied:
I can live with that.

Dreamescaper commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +26,91 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
+                       (argumentSymbol is IFieldSymbol fieldSymbol && fieldSymbol.IsConst);
+            }
+
+            bool IsEmpty(ExpressionSyntax expression)
```
```suggestion
            bool IsStringEmpty(ExpressionSyntax expression)
```

IsEmpty implies that `expression` itself is empty, so would be better to rename it somehow

manfred-brands replied:
Done.

manfred-brands commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +26,91 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
+                        .FirstOrDefault();
+            }
+
+            return null;
```
Done.

manfred-brands commented:
Added a fixup commit for the items mentioned.

mikkelbu commented on src/nunit.analyzers.tests/ConstActualValueUsage/ConstActualValueUsageAnalyzerTests.cs:
```diff
@@ -128,11 +129,37 @@ public void Test()
+            var testCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings(@"
+                public void Test()
+                {
+                    string actual = ""act"";
+                    Assert.AreEqual(actual, string.Empty);
```
```suggestion
                    string expected = ""exp"";
                    Assert.AreEqual(expected, ↓string.Empty);
```

manfred-brands replied:
No, string.Empty is the expected value. The code was supposed to test if an actual value is string.Empty. As the arguments are reversed, this should trigger an analyzer violation. String.Empty is treated as if it is a constant.

mikkelbu replied:
Ok. Then I misunderstod the intent. The signature of `Assert.AreEqual` is `Assert.AreEqual(object expected, object actual)`, so the first argument is the expected value and the second argument is the actual value - one reason to prefer to use `Assert.That(actual, Is.EqualTo(expected))` syntax.

mikkelbu commented on src/nunit.analyzers.tests/ConstActualValueUsage/ConstActualValueUsageAnalyzerTests.cs:
```diff
@@ -128,11 +129,37 @@ public void Test()
+            var testCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings(@"
+                public void Test()
+                {
+                    string actual = ""act"";
+                    Assert.That(string.Empty, Is.EqualTo(actual));
```
```suggestion
                    string expected = ""exp"";
                    Assert.That(↓string.Empty, Is.EqualTo(expected));
```

manfred-brands replied:
No, string.Empty is the expected value. The code was supposed to test if an actual value is string.Empty. As the arguments are reversed, this should trigger an analyzer violation. String.Empty is treated as if it is a constant.
One wouldn't write this code, but it could be the result of the CodeFix to convert Classic AreEqual into Constraint model.

mikkelbu replied:
Similar comment as above. The signature of `Assert.That` is `Assert.That<TActual>(TActual actual, IResolveConstraint expression)`, so the first argument is the actual value and the last is the expected value expressed as a constraint.

manfred-brands commented:
What is the ↓ character you added? I've seen it in some tests, but don't know what it is for now why.
When I extract the string to actual code the C# compiler complains about that character.

mikkelbu commented:
The `↓` character is a special character for this test framework that specifies where the diagnosis will occur. If you don't specify it the test framework will just check that the specified diagnosis will fire, but not where in the syntax it will fire.

mikkelbu commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +27,90 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
         protected override void AnalyzeAssertInvocation(SyntaxNodeAnalysisContext context,
             InvocationExpressionSyntax assertExpression, IMethodSymbol methodSymbol)
         {
+            bool IsLiteralExpression(ExpressionSyntax expression)
```
I think we should have at least one test per case in this method - so that we don't introduce regressions later on.

manfred-brands replied:
There are use cases in `AnalyzeWhenLiteralArgumentIsProvidedForAssertThat`: `-1` triggers `PrefixUnaryExpressionSyntax` and the `(2 + 3) * 1024` triggers all the others:  `ParenthesizedExpressionSyntax`, `BinaryExpressionSyntax`, `LiteralExpressionSyntax` Would you prefer these to be in separate test methods?

mikkelbu replied:
My fault, I overlooked those (so much for reviewing the tests one day and the prod code the next). I think it is fine as it is.

mikkelbu commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +27,90 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
+
+            if (expectedExpression != null && !IsConstant(expectedExpression) && !IsStringEmpty(expectedExpression))
             {
                 context.ReportDiagnostic(Diagnostic.Create(
                     descriptor,
                     actualExpression.GetLocation()));
```
```suggestion
                Report(actualExpression);
```

mikkelbu commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +27,90 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
-            var argumentSymbol = context.SemanticModel.GetSymbolInfo(actualExpression).Symbol;
+            if (IsLiteralExpression(actualExpression))
+            {
+                // Classic error case
```
I think this comment can be removed as the code is self explanatory
```suggestion
```

mikkelbu commented on src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:
```diff
@@ -24,21 +27,90 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
+            }
+
+            if (!IsConstant(actualExpression) && !IsStringEmpty(actualExpression))
+                return; // Standard use case
```
Only a suggestion
```suggestion
                return;
```
